source/common/tcp_proxy/tcp_proxy.cc

```diff
@@ -885,9 +915,24 @@ void Filter::onUpstreamEvent(Network::ConnectionEvent event) {
 
 void Filter::onUpstreamConnection() {
   connecting_ = false;
-  // Re-enable downstream reads now that the upstream connection is established
-  // so we have a place to send downstream data to.
-  read_callbacks_->connection().readDisable(false);
+
+  if (!receive_before_connect_) {
+    // Re-enable downstream reads now that the upstream connection is established
+    // so we have a place to send downstream data to.
+    read_callbacks_->connection().readDisable(false);
+  }
+  // If we have received any data before upstream connection is established, send it to
+  // the upstream connection.
+  else if (early_data_buffer_.length() > 0) {
```

Comment:
Yes it can happen that receive_before_connect is true but there was no early data so in that case we should not do anything after upstream connection is established. We should take action to flush early data buffer only if early data was received, thats why we need this else.